controllers/disruption_rollout_controller.go

```diff
@@ -0,0 +1,247 @@
+	}
+
+	if !instance.DeletionTimestamp.IsZero() {
+		// Add finalizer here if required
+		return ctrl.Result{}, nil
+	}
+
+	// Update the DisruptionRollout status based on the presence of the target resource
+	// If the target resource has been missing for longer than the TargetResourceMissingThreshold, delete the instance
+	targetResourceExists, instanceDeleted, err := r.updateTargetResourcePreviouslyMissing(ctx, instance)
+	if err != nil {
+		// Log error and requeue if status update or deletion fails
+		r.log.Errorw("failed to handle target resource status", "err", err)
+		return ctrl.Result{}, err
+	}
+
+	if instanceDeleted {
+		// Skip reconciliation since the instance has been deleted
+		return ctrl.Result{}, nil
+	}
+
+	disruptions, err := GetChildDisruptions(ctx, r.Client, r.log, instance.Namespace, DisruptionRolloutNameLabel, instance.Name)
+	if err != nil {
+		return ctrl.Result{}, nil
+	}
+
+	// Update the DisruptionRollout status with the time when the last disruption was successfully scheduled
+	if err := r.updateLastScheduleTime(ctx, instance, disruptions); err != nil {
+		r.log.Errorw("unable to update LastScheduleTime of DisruptionCron status", "err", err)
+		return ctrl.Result{}, err
+	}
+
+	// Calculate next requeue time
+	requeueAfter := time.Duration(randSource.Intn(5)+15) * time.Second //nolint:gosec
+	requeueTime := requeueAfter.Round(time.Second)
+	scheduledResult := ctrl.Result{RequeueAfter: requeueAfter}
+
+	// Run a new disruption if the following conditions are met:
+	// 1. The target resource is available
+	// 2. The target resource has been updated
+	// 3. The target resource update has not been tested
+	// 4. It's not blocked by another disruption already running
+	// 5. It's not past the deadline
+	if !targetResourceExists {
+		r.log.Infow(fmt.Sprintf("target resource is missing, scheduling next check in %s", requeueTime))
+		return scheduledResult, nil
+	}
+
+	if !r.targetResourceUpdated(&instance.Status) {
+		r.log.Infow("target resource hasn't been modified yet, sleeping")
+		return ctrl.Result{}, nil
+	}
+
+	if instance.Status.LastContainerChangeTime.Before(instance.Status.LastScheduleTime) || instance.Status.LastContainerChangeTime.Equal(instance.Status.LastScheduleTime) {
+		r.log.Infow("target resource update has already been tested, sleeping")
+		return ctrl.Result{}, nil
+	}
+
+	if len(disruptions.Items) > 0 {
+		r.log.Infow(fmt.Sprintf("cannot start a new disruption as a prior one is still running, scheduling next check in %s", requeueTime), "numActiveDisruptions", len(disruptions.Items))
+		return scheduledResult, nil
+	}
+
+	tooLate := false
+	if instance.Spec.DelayedStartTolerance.Duration() > 0 && !instance.Status.LastContainerChangeTime.IsZero() {
+		tooLate = instance.Status.LastContainerChangeTime.Add(instance.Spec.DelayedStartTolerance.Duration()).Before(time.Now())
+	}
+
+	if tooLate {
+		r.log.Infow("missed schedule to start a disruption, sleeping")
```

Comment:
Could we include LastContainerChangeTime and DelayedStartTolerance as tags in this log?